## Preprocessing the Data

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf
import os

#  Import and read the charity_data.csv.
import pandas as pd 
df = pd.read_csv("../Resources/charity_data.csv")
df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
df = df.drop("EIN", 1)
df = df.drop("NAME", 1)
df

C:\Users\Thissen\AppData\Local\Temp\ipykernel_11928\2513471742.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  df = df.drop("EIN", 1)
C:\Users\Thissen\AppData\Local\Temp\ipykernel_11928\2513471742.py:2: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  df = df.drop("NAME", 1)


APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 10 columns]

In [3]:
df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [4]:
# Look at APPLICATION_TYPE value counts for binning
#df['APPLICATION_TYPE'].value_counts()
application_counts = df['APPLICATION_TYPE'].value_counts()
application_counts

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [5]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace = list(application_counts[application_counts < 500].index)

# Replace in dataframe
for app in application_types_to_replace:
    df['APPLICATION_TYPE'] = df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [6]:
# Look at CLASSIFICATION value counts for binning
# df['CLASSIFICATION'].value_counts()
class_counts = df['CLASSIFICATION'].value_counts()
class_counts

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [7]:
# You may find it helpful to look at CLASSIFICATION value counts >1
class_count_1=class_counts[class_counts>1]
class_count_1

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [8]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classifications_to_replace = list(class_counts[class_counts < 1000].index)

# Replace in dataframe
for cls in classifications_to_replace:
    df['CLASSIFICATION'] = df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [9]:
# Convert categorical data to numeric with `pd.get_dummies`
dummies_df = pd.get_dummies(df, drop_first = True)
dummies_df

STATUS   ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_T10  \
0           1      5000              1                     1   
1           1    108590              1                     0   
2           1      5000              0                     0   
3           1      6692              1                     0   
4           1    142590              1                     0   
...       ...       ...            ...                   ...   
34294       1      5000              0                     0   
34295       1      5000              0                     0   
34296       1      5000              0                     0   
34297       1      5000              1                     0   
34298       1  36500179              0                     0   

       APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  \
0                         0                    0                    0   
1                         0                    1                    0   
2                         0                    0                    0   
3                         0                    1                    0   
4                         0                    1                    0   
...                     ...                  ...                  ...   
34294                     0                    0                    1   
34295                     0                    0                    1   
34296                     0                    1                    0   
34297                     0                    0                    0   
34298                     0                    1                    0   

       APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  APPLICATION_TYPE_T7  ...  \
0                        0                    0                    0  ...   
1                        0                    0                    0  ...   
2                        1                    0                    0  ...   
3                        0                    0                    0  ...   
4                        0                    0                    0  ...   
...                    ...                  ...                  ...  ...   
34294                    0                    0                    0  ...   
34295                    0                    0                    0  ...   
34296                    0                    0                    0  ...   
34297                    1                    0                    0  ...   
34298                    0                    0                    0  ...   

       ORGANIZATION_Trust  INCOME_AMT_1-9999  INCOME_AMT_10000-24999  \
0                       0                  0                       0   
1                       0                  1                       0   
2                       0                  0                       0   
3                       1                  0                       1   
4                       1                  0                       0   
...                   ...                ...                     ...   
34294                   0                  0                       0   
34295                   0                  0                       0   
34296                   0                  0                       0   
34297                   0                  0                       0   
34298                   0                  0                       0   

       INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  \
0                             0                   0                 0   
1                             0                   0                 0   
2                             0                   0                 0   
3                             0                   0                 0   
4                             1                   0                 0   
...                         ...                 ...               ...   
34294                         0                   0         

In [10]:
# Split our preprocessed data into our features and target arrays
X = dummies_df.drop(["IS_SUCCESSFUL"], axis='columns')#.values
y = dummies_df["IS_SUCCESSFUL"]#.values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [11]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [12]:
inputs = len(X.columns)

In [13]:
print(inputs)

36


## Compile, Train and Evaluate the Model - Second Attempt

In [14]:
# Create a method that creates a new Sequential model with hyperparameter options
def create_model(hp):
    nn_model = tf.keras.models.Sequential()

    # Allow kerastuner to decide which activation function to use in hidden layers
    activation = hp.Choice('activation',['relu','tanh'])
    
    # Allow kerastuner to decide number of neurons in first layer
    nn_model.add(tf.keras.layers.Dense(units=hp.Int('first_units',
        min_value=6,
        max_value=18,
        step=6), activation=activation, input_dim=inputs))

    # Allow kerastuner to decide number of hidden layers and neurons in hidden layers
    for i in range(hp.Int('num_layers', 1, 2)):
        nn_model.add(tf.keras.layers.Dense(units=hp.Int('units_' + str(i),
            min_value=6,
            max_value=12,
            step=6),
            activation=activation))
    
    nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

    # Compile the model
    nn_model.compile(loss="binary_crossentropy", optimizer='adam', metrics=["accuracy"])
    
    return nn_model

In [15]:
# Import the kerastuner library
import keras_tuner as kt

tuner = kt.Hyperband(
    create_model,
    objective="val_accuracy",
    max_epochs=20,
    hyperband_iterations=2,
directory=os.path.normpath('C:/')
)

In [16]:
# Run the kerastuner search for best hyperparameters
tuner.search(X_train_scaled,y_train,epochs=20,validation_data=(X_test_scaled,y_test))

Trial 60 Complete [00h 00m 43s]
val_accuracy: 0.727580189704895

Best val_accuracy So Far: 0.7280466556549072
Total elapsed time: 00h 16m 11s
INFO:tensorflow:Oracle triggered exit


## Compile, Train and Evaluate the Model - Third Attempt

In [21]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

nn_model = tf.keras.models.Sequential()

# First hidden layer
nn_model.add(tf.keras.layers.Dense(units=18, activation="relu", input_dim=36))

# Second hidden layer
nn_model.add(tf.keras.layers.Dense(units=18, activation="sigmoid"))

# Output layer
nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_model.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 18)                666       
                                                                 
 dense_4 (Dense)             (None, 18)                342       
                                                                 
 dense_5 (Dense)             (None, 1)                 19        
                                                                 
Total params: 1,027
Trainable params: 1,027
Non-trainable params: 0
_________________________________________________________________


In [22]:
# Compile the model
nn_model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [23]:
# Train the model
from keras.callbacks import Callback

mc = tf.keras.callbacks.ModelCheckpoint('weights{epoch:08d}.h5', 
                                     save_weights_only=True, period=5)
    
        
fit_model = nn_model.fit(X_train_scaled, y_train, epochs=30, callbacks = [mc])

Epoch 1/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5987 - accuracy: 0.6984
Epoch 2/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5626 - accuracy: 0.7280
Epoch 3/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5574 - accuracy: 0.7311
Epoch 4/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5550 - accuracy: 0.7298
Epoch 5/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5535 - accuracy: 0.7322
Epoch 6/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5514 - accuracy: 0.7322
Epoch 7/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5511 - accuracy: 0.7321
Epoch 8/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5502 - accuracy: 0.7325
Epoch 9/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5496 - accuracy: 0.7337
Epoch 10/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5496 - accuracy: 0.7328

In [24]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5534 - accuracy: 0.7252 - 297ms/epoch - 1ms/step
Loss: 0.5534136891365051, Accuracy: 0.7252478003501892


## Compile, Train and Evaluate the Model - Fourth Attempt

In [ ]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

nn_model = tf.keras.models.Sequential()

# First hidden layer
nn_model.add(tf.keras.layers.Dense(units=18, activation="relu", input_dim=36))

# Second hidden layer
nn_model.add(tf.keras.layers.Dense(units=18, activation="tanh"))

# Output layer
nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_model.summary()

In [25]:
# Compile the model
nn_model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [26]:
# Train the model
from keras.callbacks import Callback

mc = tf.keras.callbacks.ModelCheckpoint('weights{epoch:08d}.h5', 
                                     save_weights_only=True, period=5)
    
        
fit_model = nn_model.fit(X_train_scaled, y_train, epochs=30, callbacks = [mc])

Epoch 1/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5455 - accuracy: 0.7345
Epoch 2/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5446 - accuracy: 0.7358
Epoch 3/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5447 - accuracy: 0.7357
Epoch 4/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5449 - accuracy: 0.7343
Epoch 5/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5447 - accuracy: 0.7357
Epoch 6/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5443 - accuracy: 0.7365
Epoch 7/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5440 - accuracy: 0.7363
Epoch 8/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5444 - accuracy: 0.7356
Epoch 9/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5439 - accuracy: 0.7355
Epoch 10/30
804/804 [==============================] - 1s 1ms/step - loss: 0.5443 - accuracy: 0.7359

In [27]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5522 - accuracy: 0.7273 - 318ms/epoch - 1ms/step
Loss: 0.5521854162216187, Accuracy: 0.7273469567298889


In [28]:
# Export our model to HDF5 file
nn_model.save("AlphabetSoupCharity_Optimization.h5")

## Analysis Overview

The purpose of this analysis is to determine an algorithm which will predict success for the nonprofit organizations applying to Alphabet Soup Charity for funding. 

## Results:

## Data Preprocessing:
- The target variable for the model is if the charities are effectively using their money received from Alphabet Soup, and this is seen in the "IS_SUCCESSFUL" column.
- The variables considered to be the features of the model are the type of application, (e.g T2, T9, T19, etc.) and classification according to type of government organization (e.g. C1000, C2100, C4500, etc.)
- The EIN and names of the organizations are not considered the target or features of the model, and therefore was dropped from the data frame. 

## Compiling, Training, and Evaluating the Model:
- For the initial model, 18 neurons were chosen for each of the first and second layers due to the dimensionality of the data (this being twice the number of the binned application types and three times the number of classification types.)
- Two hidden layers were chosen for each of the features of the model. Relu was chosen as the activation function for the initial two hidden layers. Since Relu does not need to perform exponential operations, use of this activation model is quicker and more reliable than either sigmoid or tanh. Sigmoid was used in the output layer to draw a distinction between two alternatives (whether or not the charity will be successful in its use of funding).
- Over four attempts to optimize the model, I was unable to meet the 75% target predictive accuracy. 
- During the second attempt, I used the Keras Tuner model to let the computer pick the best number of neurons and layers as well as activation functions. The Keras Tuner model yielded an accuracy, when rounded, is identical to the initial accuracy of 0.728. 
- In the third attempt, I modified the activation model of the layer to sigmoid, and got a slightly reduced accuracy of 0.725. 
- For the final attempt, I decided to further pursue the activation models and designate the first layer as relu, but the second layer as tanh, and the third layer remained sigmoid. I left the number of neurons the same so that no other variables would explain the accuracy obtained. The fourth and final accuracy obtained was 0.727, slightly better than the third attempt but still just under the initial and second model accuracy of 0.728. 

## Summary:

- The best results for the models were yielded with each hidden layer being a relu function, and allowing the Keras Tuner to select neurons, layers and activation functions. 
- For further models, I would add more neurons to each layer, more hidden layers based on the number of features, and vary the number of epochs over which the model is trained. 
- Additional features could include further binning of the application types and classification numbers. In addition, the amount of times a charity has received funding could be used as an additional feature. 
- A recommendation for another model to use in this situation would be the Random Forest model. This model is good for classification problems because there is a robust number of decision trees. It is also not affected by overfitting. 